### PR TITLE
Calculate displacement operator unitary matrix using The Walrus

### PR DESCRIPTION
**Pull request by sduquemesa**
**Context:**
[The Walrus PR #351](https://github.com/XanaduAI/thewalrus/pull/351) implemented a more stable calculation of the displacement unitary.

**Description of the Change:**
The `Dgate` now uses The Walrus to calculate the unitary and gradients of the displacement gate.

Before:
before

After:
after

**Benefits:**
This provides better numerical stability for larger cutoff and displacement values.

**Possible Drawbacks:**
None

**Related GitHub Issues:**
None

### CODE REVIEW

File: mrmustard/lab/gates.py [hunk 1]
```diff
@@ -93,6 +93,48 @@ def __init__(
...
+
+        r = (
+            math.sqrt(self.x.value**2 + self.y.value**2)
+            if N > 1
+            else math.sqrt([self.x.value**2 + self.y.value**2])
+        )
+        phi = (
+            math.atan(self.y.value / self.x.value)
+            if N > 1
+            else math.atan([self.y.value / self.x.value])
```
Comment (ziofil) [inline mrmustard/lab/gates.py:114]: let's use atan2 instead

File: mrmustard/math/math_interface.py [hunk 2]
```diff
@@ -252,6 +252,17 @@ def cosh(self, array: Tensor) -> Tensor:
             array: hyperbolic cosine of array
         """
 
+    @abstractmethod
+    def atan(self, array: Tensor) -> Tensor:
```
Comment (ziofil) [inline mrmustard/math/math_interface.py:256]: switch to atan2

**Review by ziofil — APPROVED** [review]
very good! Just let's change atan to atan2 (it's just better and we capture all the cases)

**Comment by sduquemesa** [comment]
This won't have any effect unless a release of tw is made 🤔

Should we wait for that before merging?

**Review by elib20 — APPROVED** [review]
Nice!!

**Pull request merged**